Update script API to take example inputs

nikithamalgifb opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56029 Add tests for PDT in JIT
* #55379 Infer argument types for functions in JIT
* **#55376 Update script API to take example inputs**

Differential Revision: [D27897350](https://our.internmc.facebook.com/intern/diff/D27897350)

nikithamalgifb committed:
Update script API to take example inputs

[ghstack-poisoned]

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -841,7 +849,7 @@ def call_prepare_scriptable_func(obj):
     return call_prepare_scriptable_func_impl(obj, memo)
 
-def script(obj, optimize=None, _frames_up=0, _rcb=None):
+def script(obj, optimize=None, _frames_up=0, _rcb=None, example_inputs: Optional[List[Tuple]]=None):
```
As discussed earlier, let's avoid publishing it to `torch.jit.script` immediately. We should test against `torch.jit._script` during development. In the future, we can merge `_script` with `script`.

And, I suggest we rename `_script` to something like `_script_pdt` so that no one accidentally uses it.

nikithamalgifb replied:
Fixed now

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -36,6 +36,14 @@
 from torch.overrides import (
     has_torch_function, has_torch_function_unary, has_torch_function_variadic)
 
+_IS_MONKEYTYPE_INSTALLED = True
```
The import guard can be avoided if we do this:
https://github.com/pytorch/pytorch/pull/54513/files#r609120259

nikithamalgifb replied:
Fixed now

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -36,6 +36,14 @@
+try:
+    from monkeytype import trace as monkeytype_trace  # type: ignore[import]
+    from torch.jit._monkey_config_jit import JitTypeTraceConfig, type_trace_db
+    monkeytype_config = JitTypeTraceConfig(type_trace_db)
```
This is related to:
https://github.com/pytorch/pytorch/pull/54513/files#r610384224

I think each `torch.jit.script` call should maintain its own trace database, sharing it among multiple calls can lead to confusing behaviors.

nikithamalgifb replied:
Agreed and fixed

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -36,6 +36,14 @@
+    monkeytype_trace,
+    JitTypeTraceConfig ,
+    JitTypeTraceStore
+)  # type: ignore[import]
```
Why is `# type: ignore` needed here?

nikithamalgifb replied:
Now removed.

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -36,6 +36,14 @@
+    JitTypeTraceStore
+)  # type: ignore[import]
+
+type_trace_db = JitTypeTraceStore()  # DB to hold all call traces from MonkeyType
```
Since each call to `_script_pdt` creates a new instance, we probably don't need to initialize here.

nikithamalgifb replied:
When monkeytype is not installed, then we need this to be initialized to be used in `torch/jit/frontend.py`

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -841,6 +849,93 @@ def call_prepare_scriptable_func(obj):
+
+        # Enums are automatically usable in TorchScript, explicitly scripting
+        # is not necessary, but not harmful either.
+        if issubclass(obj, enum.Enum):
+            return obj
+
+        if not _is_new_style_class(obj):
+            raise RuntimeError(
+                "TorchScript classes must be new-style classes. "
+                "Please inherit from 'object'."
+            )
+        if len(obj.mro()) > 2:
+            raise RuntimeError(
+                "TorchScript classes does not support inheritance yet. "
+                "Please directly inherit from 'object'."
+            )
+        if _rcb is None:
+            _rcb = _jit_internal.createResolutionCallbackFromFrame(_frames_up + 1)
+        _compile_and_register_class(obj, _rcb, qualified_name)
+        return obj
+    else:
+        # this is a decorated fn, and we need to the underlying fn and its rcb
+        if hasattr(obj, "__script_if_tracing_wrapper"):
+            obj = obj.__original_fn
+            _rcb = _jit_internal.createResolutionCallbackFromClosure(obj)
+
+        _check_directly_compile_overloaded(obj)
+        maybe_already_compiled_fn = _try_get_jit_cached_function(obj)
+        if maybe_already_compiled_fn:
+            return maybe_already_compiled_fn
+        ast = get_jit_def(obj, obj.__name__)
+        if _rcb is None:
+            _rcb = _jit_internal.createResolutionCallbackFromClosure(obj)
+        fn = torch._C._jit_script_compile(
+            qualified_name, ast, _rcb, get_default_args(obj)
+        )
+        # Forward docstrings
+        fn.__doc__ = obj.__doc__
+        _set_jit_function_cache(obj, fn)
+        return fn
```
Can we call `script` here instead of duplicating code? I guess this is because we want to run some sanity checks before calling MonkeyType, but it is probably less important during prototyping.

nikithamalgifb replied:
Ok yeah, for now we can do that. Changed

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -841,6 +849,93 @@ def call_prepare_scriptable_func(obj):
+def _script_pdt(obj, optimize=None, _frames_up=0, _rcb=None, example_inputs: Optional[List[Tuple]] = None):
+    # This is a private API, intended for internal use only . Usage of this API is only for experimental
+    # purposes only and is highly discouraged to be used.
+    global type_trace_db
```
TODO for future:
We should avoid using global variable for `type_trace_db` because this allows concurrent scripting to interfere with each other, but this is not critical for prototyping.

nikithamalgifb replied:
Agreed!

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -840,7 +848,45 @@ def call_prepare_scriptable_func(obj):
+    if isinstance(obj, ScriptFunction):
+        return obj
+
+    if isinstance(obj, torch.nn.Module):
+        obj = call_prepare_scriptable_func(obj)
+        return torch.jit._recursive.create_script_module(
+            obj, torch.jit._recursive.infer_methods_to_compile
+        )
+
```
This skips `nn.Module` compilation I think. Is this intended?

nikithamalgifb replied:
I think it was an oversight from my end. Now fixed

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -106,6 +114,9 @@ def __init__(self):
 """
 
+def _get_type_trace_db():
+    # This is a private API. Use of this for external puposes is discouraged.
```
```suggestion
    # This is a private API. Use of this for external purposes is discouraged.
```

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -840,7 +851,43 @@ def call_prepare_scriptable_func(obj):
 
-def script(obj, optimize=None, _frames_up: int = 0, _rcb=None):
+def _script_pdt(obj, optimize=None, _frames_up=0, _rcb=None, example_inputs: Optional[List[Tuple]] = None):
+    # This is a private API, intended for internal use only . Usage of this API is only for experimental
+    # purposes only and is highly discouraged to be used.
```
```suggestion
    # This is a private API, intended for internal use only. Usage of this API is only for experimental
    # purposes only and is highly discouraged.
```

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -840,7 +851,43 @@ def call_prepare_scriptable_func(obj):
+                obj(*example_input)
+    else:
+        warnings.warn("Warning: monkeytype is not installed. Please install https://github.com/Instagram/MonkeyType "
+                      "to enable Profile-Directed Typing in TorchScript. Refer  "
+                      "https://github.com/Instagram/MonkeyType/blob/master/README.rst to install MonkeyType. ")
```
```suggestion
                      "to enable Profile-Directed Typing in TorchScript. Refer to"
                      "https://github.com/Instagram/MonkeyType/blob/master/README.rst to install MonkeyType. ")
```

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"




[ghstack-poisoned]

Commit:
Update on "Update script API to take example inputs"


Differential Revision: [D27897350](https://our.internmc.facebook.com/intern/diff/D27897350)

[ghstack-poisoned]
